src/core/qgstiles.cpp

```diff
@@ -16,22 +16,65 @@
 #include "qgstiles.h"
 
 #include "qgslogger.h"
+#include "qgscoordinatereferencesystem.h"
+#include "qgssettings.h"
 
-QgsTileMatrix QgsTileMatrix::fromWebMercator( int zoomLevel )
+QgsTileMatrix QgsTileMatrix::fromWebMercator( const int &zoomLevel )
 {
+  double z0xMin = -20037508.3427892, z0yMax =  20037508.3427892;
+
+  return fromCustomDef( zoomLevel, QgsCoordinateReferenceSystem( "EPSG:3857" ), QgsPointXY( z0xMin, z0yMax ), 2 * z0yMax );
+}
+
+QgsTileMatrix QgsTileMatrix::fromCustomDef( const int &zoomLevel, const QgsCoordinateReferenceSystem &crs,
+    const QgsPointXY &z0TopLeftPoint, const double &z0Dimension,
+    const int &z0MatrixWidth, const int &z0MatrixHeight )
+{
+  // Square extent calculation
+  double z0xMin = z0TopLeftPoint.x();
+  double z0yMax = z0TopLeftPoint.y();
+  double z0xMax = z0xMin + z0MatrixWidth * z0Dimension;
+  double z0yMin = z0yMax - z0MatrixHeight * z0Dimension;
+
+  // Constant for scale denominator calculation
+  const double tileSize = 256.0;
+  const double POINTS_TO_M = 2.8 / 10000.0; //2.800005600011068 / 10000.0;
```

Comment:
I would suggest renaming the variable and adding a comment for the future reference
```suggestion
  const double PIXELS_TO_M = 0.28 / 1000.0;  // WMS/WMTS define "standardized rendering pixel size" as 0.28mm
```